### ISSUE TITLE
Problem with lzfse_decode_buffer_with_scratch

### ISSUE THREAD

**Issue by kleuter** [lzfse/lzfse#44]
When checking return value of **lzfse_decode_buffer** (which calls **lzfse_decode_buffer_with_scratch**), I'm supposed to treat **dst_size** as error:

```C++
if (status == LZFSE_STATUS_DST_FULL)
    return dst_size;
```

But I've encountered a situation, when the last line of **lzfse_decode_buffer_with_scratch**:
```C++
return (size_t)(s->dst - dst_buffer); // bytes written
```
also returns **dst_size**! Which must not be treated as error.

The current workaround is returning **0** instead of **dst_size**, but it can be a special handling e.g. extra error parameter.

**Comment by shampoofactory** [comment]
> Is this also related to #44 ?

TLDR: It's related but does not solve the issue because there's an unhandled corner case.

I withdrew that commit as there is an unhandled corner case (*). Sorry.

I'm not sure I have any good answers for you but here are my two cents. Most of which I suspect you already know.

Within the context of `lzfse_decode_buffer` my understanding is that the return value, let's call it `ret`, can communicate three outcomes:
 
1. Success. Here `ret` reflects the number of bytes written into the destination buffer.
2. Destination buffer is too small. Here `ret` equals the size of the destination buffer.
3. Decode error. Here `ret` equals 0. This is not explicitly documented but we can see the relevant code [here](https://github.com/lzfse/lzfse/blob/e634ca58b4821d9f3d560cdc6df5dec02ffc93fd/src/lzfse_decode.c#L47).

The issue is that `ret` is overloaded. It's not always possible to assign an outcome based on `ret` alone.

For example, let's take a `ret` value of zero. This could mean:

1. Success. Where we've decoded a zero-length input file.
2. Destination buffer is too small. Where we've passed a zero-sized destination buffer.
3. Decode error. Where we've passed invalid data.

Although we can exclude scenario 2 by not passing in zero-sized buffers, we still cannot differentiate between scenarios 1 and 3 (*). We need more information. At some level, we need to tweak the API. One such suggestion is [here](https://github.com/lzfse/lzfse/issues/5#issuecomment-237134992).
